Image viewer 2d projection

toastisme opened the pull request:
This allows 2d projections to be added to a Detector at the creation stage. 

The purpose of this is to avoid special case checks buried in image_viewer scripts for detectors that do not work with project_2d (i.e FormatCBFMiniPilatusDLS12M). Unique projection behaviour is now defined only in a format class, making it much easier to add more edge case detectors. image_viewer scripts then use explicit projections over project_2d if they are present on the Detector. projection_2d keys are only added to .expt files if they have been manually set, to avoid noise in the majority of cases with normal detectors.

toastisme commented:
@rjgildea suggested ensuring the panel projection 2d getter is consistent with what comes out of `project_2d`. I've renamed `project_2d` to `get_detector_projection_2d_axes`, to better describe what it does. I've then abstracted out the equivalent for `panel.get_projection_2d` from `tile_generation.py` into `detector_helpers.py` as `get_panel_projection_2d_from_axes`. This function returns rotation and translation, like `panel.get_projection_2d`.

dagewa commented on newsfragments/422.misc:
```diff
@@ -0,0 +1,2 @@
+2D projection getters/setters added to Panel.
+Panel 2D projection using axes moved from DIALS to detector_helpers.py and made consistent with panel.get_2d_projection.
```
I defer to @ndevenish for comment

dagewa commented on src/dxtbx/model/detector_helpers.py:
```diff
@@ -11,6 +11,8 @@
 except ImportError:
     sklearn = None
 
+from xfel.cftbx.detector.metrology import get_projection_matrix
```
We should avoid a dependence on `xfel` if possible as `dxtbx` should remain an upstream dependency for DIALS and cctbx.xfel. That function looks fairly basic so perhaps could just be reproduced here, or the original moved to e.g. `scitbx`

dagewa commented on src/dxtbx/model/detector_helpers.py:
```diff
@@ -294,9 +296,15 @@ def set_detector_distance(detector, distance):
-def project_2d(detector):
-    """
-    Project panel origin, fast and slow onto the best-fitting 2D plane.
+def get_detector_projection_2d_axes(detector):
```
:+1:

dagewa reviewed (changes requested):
I tested with an I23 image and that seemed fine. Appreciate the campsite tidying and naming improvements. The main issue is the addition of a dxtbx→xfel dependency

toastisme commented:
Removed xfel dependency. As it's one line (and we're not even using all of the output of get_projection_matrix) I've just added it explicitly. If get_projection_matrix was moved, there's probably a conversation about having it in numpy instead, where it should be, and making changes to it being used in cctbx, which feels out of scope for this PR.

rjgildea commented on tests/model/test_detector.py:
```diff
@@ -305,11 +309,135 @@ def test_project_2d():
+    detector = create_detector(offset=0)
+
+    # Test detector with no projections set explicity has no 2d projection
+    assert detector.has_projection_2d() is False
```
See comments above r.e. explicitly comparing to bool.
```suggestion
    assert not detector.has_projection_2d()
```

rjgildea commented on src/dxtbx/model/detector.h:
```diff
@@ -763,6 +763,15 @@ namespace dxtbx { namespace model {
         (*this)[i].rotate_around_origin(axis, angle);
       }
     }
+  
+    bool has_projection_2d(){
+      for (std::size_t i = 0; i < size(); ++i){
+        if (!(*this)[i].has_projection_2d()){
+          return false;
+        }
+      }
+      return true;
+    }
```
Do we need an explicit `Detector.has_projection_2d()` function, particularly if we're making the suggested changes below to remove `Panel.has_projection_2d()` and allow `Panel.get_projection_2d()` to optionally return `None`. Then you would just be able to do e.g. `all([p.get_projection_2d() for p in detector])` in Python, but with more control over what to do if only _some_ of the panels have a 2d projection.

rjgildea commented on src/dxtbx/model/detector_helpers.py:
```diff
@@ -294,9 +294,15 @@ def set_detector_distance(detector, distance):
+
+    """Project panel origins, fast and slow axes onto the best-fitting 2D plane.
+
+    :param detector: detector used to generate axes
+    :type detector: Detector
+
+    :return: 2D origin, fast, and slow vectors for each panel
+    :rtype: list(tuple(float, float)), list(tuple(float, float)), list(tuple(float, float))
```
As below, I think type annotations would be more modern/useful here than this style of docstrings.

rjgildea reviewed (changes requested):
Thank you for adding this, it will be a good addition to remove hardcoded cases for the P12M in the dials.image_viewer, as well as enabling support for similarly non-flat detectors.

I think the C++ interface could potentially be cleaned up and made more "Pythonic" by allowing the projections to be `None` using `boost::optional`, and it would probably be more useful to use type annotations rather than docstrings to indicate parameter types (although I appreciate this is still better than nothing at all).

toastisme commented:
@rjgildea thanks for the detailed feedback. Had not come across boost::optional before. I think I've applied all the suggestions. The only thing I've left is the `Detector.has_projection_2d()`. It's largely syntactic sugar but I also included it to avoid partially set projections. Is there a use case for setting some panels manually, and others using the beam centre (@dagewa)? That feels like it would be a mess to me, but happy to change it if it's viable.

rjgildea commented on src/dxtbx/model/detector_helpers.py:
```diff
@@ -294,15 +300,10 @@ def set_detector_distance(detector, distance):
-
-    :param detector: detector used to generate axes
-    :type detector: Detector
+def get_detector_projection_2d_axes(detector: Detector) -> Tuple[Tuple, Tuple, Tuple]:
```
You could define a [type alias](https://docs.python.org/3/library/typing.html#type-aliases) to make the return signature slightly clearer:
```suggestion
Coord2d = Tuple[float, float]

def get_detector_projection_2d_axes(detector: Detector) -> Tuple[Coord2d, Coord2d, Coord2d]:
```

rjgildea commented on src/dxtbx/model/detector_helpers.py:
```diff
@@ -410,25 +411,16 @@ def get_detector_projection_2d_axes(detector):
+    fast_axis_2d: matrix.col,
+    slow_axis_2d: matrix.col,
+    origin_2d: matrix.col,
+) -> Tuple[Tuple, Tuple]:
```
As above, perhaps better naming could be something like:
```
Float2 = Tuple[float, float]
Float4 = Tuple[float, float, float, float]

def get_panel_projection_2d_from_axes(...) -> Tuple[Float4, Float2]:
    ...
```

dagewa commented:
> @rjgildea thanks for the detailed feedback. Had not come across `boost::optional` before. I think I've applied all the suggestions. The only thing I've left is the `Detector.has_projection_2d()`. It's largely syntactic sugar but I also included it to avoid partially set projections. Is there a use case for setting some panels manually, and others using the beam centre (@dagewa)? That feels like it would be a mess to me, but happy to change it if it's viable.

I can't think of a reason to only set some panels. When we're doing this, we want to do it to the whole detector
